### PR TITLE
Add fuse-cast-initializer pass that fuses a cast operation that is applied to an initializer into the initializer

### PR DESCRIPTION
**Pull request by jfrohnhofen**
Before:
A is in the initializer list
X = Cast(A)
Z = X + Y

After:
B is in the initializer list
Z = B + Y

This PR also adds a SETUP.go file that builds the onnx-optimizaer library so that it can be used from Go as a DBT dependency

### CODE REVIEW

File: onnxoptimizer/passes/fuse_cast_into_initializer.h [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+void copyData(std::vector<A>& dst, const B* src, const size_t len) {
+    dst.reserve(len);
+    for (size_t i=0; i<len; ++i) {
+        dst.push_back(static_cast<T>(src[i]));
+    }
+}
+
+}  // namespace
+
+struct FuseCastIntoInitializer final : public PredicateBasedPass {
```
Comment (ljanyst) [inline onnxoptimizer/passes/fuse_cast_into_initializer.h:37]: Can you please briefly document the behavior in the comment?

File: onnxoptimizer/passes/fuse_cast_into_initializer.h [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  explicit FuseCastIntoInitializer()
+      : PredicateBasedPass(PassType::Fuse, PassEfficiency::Partial,
+                           PassOptimizationType::Compute) {}
+
+  std::string getPassName() const override {
+    return "fuse_cast_into_initializer";
+  }
+
+  bool patternMatchPredicate(Node *node) override {
+    return node->kind() == kCast && node->input()->node()->kind() == kParam && node->input()->elemType() != node->i(kto);
```
Comment (ljanyst) [inline onnxoptimizer/passes/fuse_cast_into_initializer.h:50]: What is `kto` and where does it come from? What is a 'kParam` node.

File: onnxoptimizer/passes/fuse_cast_into_initializer.h [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+//   B is in the initializer list
+//   Z = B + Y
+//
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+namespace {
+
+enum DataType {
```
Comment (ljanyst) [inline onnxoptimizer/passes/fuse_cast_into_initializer.h]: They really don't have their own enum for that?

File: onnxoptimizer/passes/fuse_cast_into_initializer.h [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  COMPLEX64 = 14,   // not supported
+  COMPLEX128 = 15,  // not supported
+  BFLOAT16 = 16     // not supported
+};
+
+template <typename T, typename A, typename B>
+void copyData(std::vector<A>& dst, const B* src, const size_t len) {
+    dst.reserve(len);
+    for (size_t i=0; i<len; ++i) {
+        dst.push_back(static_cast<T>(src[i]));
```
Comment (ljanyst) [inline onnxoptimizer/passes/fuse_cast_into_initializer.h:31]: That's just truncating in the case of ints. Wouldn't normalization be better?

**Review by ljanyst — APPROVED** [review]

File: onnxoptimizer/passes/fuse_cast_into_initializer.h (hunk 4 above)
Comment (jfrohnhofen) [reply onnxoptimizer/passes/fuse_cast_into_initializer.h:31]: Maybe better, but that is not how the ONNX Cast operator is defined.

File: onnxoptimizer/passes/fuse_cast_into_initializer.h (hunk 1 above)
Comment (jfrohnhofen) [reply onnxoptimizer/passes/fuse_cast_into_initializer.h:37]: Added comment

File: onnxoptimizer/passes/fuse_cast_into_initializer.h (hunk 2 above)
Comment (jfrohnhofen) [reply onnxoptimizer/passes/fuse_cast_into_initializer.h:50]: kto is the constant for the name of the "to" attribute that specifies the destination type.
kParam is a node that is an initializer.
Added a comment that explains both

File: onnxoptimizer/passes/fuse_cast_into_initializer.h (hunk 3 above)
Comment (jfrohnhofen) [reply onnxoptimizer/passes/fuse_cast_into_initializer.h]: They do. Updates.

**Pull request merged**